x-pack/legacy/plugins/index_management/public/app/components/mappings_editor/lib/search_fields.tsx

```diff
@@ -0,0 +1,227 @@
+    score += 5;
+  }
+
+  return score;
+};
+
+const getJSXdisplayFromMeta = (
+  searchData: SearchData,
+  fieldData: FieldData,
+  metadata: Omit<SearchMetadata, 'score' | 'display'>
+): JSX.Element => {
+  const { term } = searchData;
+  const { path } = fieldData;
+
+  let display: JSX.Element = <span>{path}</span>;
+
+  if (metadata.fullyMatchPath) {
+    display = (
+      <span>
+        <strong>{path}</strong>
+      </span>
+    );
+  } else if (metadata.matchStartOfPath) {
+    const endString = path.substr(term.length, path.length);
+    display = (
+      <span>
+        <strong>{term}</strong>
+        {endString}
+      </span>
+    );
+  } else if (metadata.matchPath) {
+    const { stringMatch } = metadata;
+    const charIndex = path.toLowerCase().indexOf(stringMatch!.toLowerCase());
+    const startString = path.substr(0, charIndex);
+    const endString = path.substr(charIndex + stringMatch!.length);
+    display = (
+      <span>
+        {startString}
+        <strong>{stringMatch}</strong>
+        {endString}
+      </span>
+    );
+  }
+
+  return display;
+};
+
+const getSearchMetadata = (searchData: SearchData, fieldData: FieldData): SearchMetadata => {
+  const { term, type, searchRegexArray } = searchData;
+  const typeToCompare = type ?? term;
+
+  const matchStartOfPath = fieldData.path.startsWith(term);
+  const fullyMatchPath = term === fieldData.path;
+  const matchType = fieldData.type.includes(typeToCompare);
+  const fullyMatchType = typeToCompare === fieldData.type;
+
+  let stringMatch: string | null = null;
+
+  if (fullyMatchPath) {
+    stringMatch = fieldData.path;
+  } else {
+    // Execute all the regEx and sort them with the one that has the most
+    // characters match first.
+    const arrayMatch = searchRegexArray
+      .map(regex => regex.exec(fieldData.path))
+      .filter(Boolean)
+      .sort((a, b) => b![0].length - a![0].length);
+
+    if (arrayMatch.length) {
+      stringMatch = arrayMatch[0]![0];
+    }
+  }
+
+  const matchPath = stringMatch !== null;
+
+  const metadata = {
+    matchPath,
+    matchStartOfPath,
+    fullyMatchPath,
+    matchType,
+    fullyMatchType,
+    stringMatch,
+  };
+
+  const score = calculateScore(metadata);
+  const display = getJSXdisplayFromMeta(searchData, fieldData, metadata);
+
+  return {
+    ...metadata,
+    display,
+    score,
+  };
+};
+
+/**
+ * Return an array of array combining sibling elements
+ * In: ['A', 'B', 'C', 'D']
+ * Out: [['A', 'B'], ['B', 'C'], ['C', 'D']]
+ *
+ * @param arr Array of string
+ */
+const getSubArrays = (arr: string[]): string[][] => {
+  let i = 0;
+  const result = [];
+  while (i < arr.length - 1) {
+    result.push([arr[i], arr[i + 1]]);
+    i++;
+  }
+  return result;
+};
+
+const getRegexFromArray = (array: string[]): RegExp[] => {
+  const termsRegexArray = array.map(value => new RegExp(value, 'i'));
+  const fuzzyJoinChar = '[_\\.-\\s]?';
+  const fuzzySearchRegexArray = getSubArrays(array).map(
```

Comment:
It looks like what this section of the search does is:

Take inputs like `[ "some", "terms", "here" ]` then returns `[ [ "some", "terms" ],  [ "terms", "here" ] ]`. These are converted into regex like `/some[_\.-\s]?terms/` but this does not preserve the fuzzy search behaviour of implementations like in VSCode x-file search (activated, on mac, with the `cmd+p` shortcut). Where one can enter `someterms` and it would match a file name + path like `/test/data/some/another/terms`.

In the example of this data:

https://gist.githubusercontent.com/cjcenizal/09837195a7b1f2ba1447aa4d67859497/raw/c8a9e541d3797adea2a6e89327f860a8263be5ab/metricbeat_index_template

I would have expected a term like `aerospiken` to match

<img width="451" alt="Screenshot 2020-01-08 at 15 22 29" src="https://user-images.githubusercontent.com/8155004/71985435-e3cb7b00-322a-11ea-8697-5662131e9898.png">


whereas, in the current implementation, it does not match anything. I think building an ever long, single, RegExp as the user enters data will exponentially increase the specificity of the term provided (for each char added) and so dramatically decrease the size of the search pool.

For example, change L125 to:
```ts
    const result = fieldData.path.match(new RegExp(term.split('').join('.*?')));
    if (result) {
      stringMatch = result[0];
    }
```

yielded this behaviour (sorting is totally wrong of course :joy:)

<img width="1264" alt="Screenshot 2020-01-08 at 15 31 48" src="https://user-images.githubusercontent.com/8155004/71986031-04e09b80-322c-11ea-968b-7360ae8196fe.png">

In combination with this, I would say that the search term box should specifically only match path and then we can filter by type in a separate area (since all types are well enumerated).

Reply:
I really like these ideas.

Reply:
Thanks for the suggestion @jloleysens !

Although this super fuzzy regex brings its own problems too 😊 as you can see here

<img width="825" alt="Screen Shot 2020-01-09 at 15 19 02" src="https://user-images.githubusercontent.com/2854616/72056964-9f57e200-32f3-11ea-894f-d03b57c098ee.png">

This is the result with the current algorithm, which IMO is more accurate

<img width="828" alt="Screen Shot 2020-01-09 at 15 23 25" src="https://user-images.githubusercontent.com/2854616/72057239-11c8c200-32f4-11ea-958f-a0bfcc8f963a.png">

Welcome to Elasticsearch engineers problems! 😄 

Like I mentioned in the PR description, I will try to improve the RegEx to get the best result possible. But we might hit a wall and if we need the precision of a VSCode, use a (tiny) library like flexsearch (especially to not re-invent the wheel).

Reply:
> Although this super fuzzy regex brings its own problems too 😊 as you can see here

I think once we sort results shortest to longest (in path length) we will get a UX very close to what VS Code has :) .

I'm not convinced we need a more sophisticated search tool here (like a library). I think we will very quickly see diminishing returns for more sophisticated tools for search. An imperative version of the algo I have in mind something like this:

https://github.com/talmobi/node-fzf/blob/4ee7fbc1d90ef61e7d04b423dff4faac2bd969a5/src/index.js#L214

> ...will realize that adding the last "n" empties the results

Hmm, I think this is probably true in the current implementation. I guess we need to decide which experience would be more/less effort for users to learn/intuit and how we present the search box. Because if they tried `aerospiken` and got no results we should not present this as fuzzy search IMO.